File: algorithms/451-sort-characters-by-frequency/sort-characters-by-frequency.py
"""
451. sort characters by frequency

https://leetcode.com/problems/sort-characters-by-frequency/
https://leetcode-cn.com/problems/sort-characters-by-frequency/

根据字符出现频率排序
"""

import collections


def frequency_sort_v1(s: str) -> str:
    """
    按照出现的频率排序
    """
    counter = collections.Counter(s)
    keys = list(counter.keys())
    keys.sort(key=lambda x: counter[x], reverse=True)
    ans = ""
    for key in keys:
        ans = ans + key * counter[key]

    return ans


def frequency_sort_v2(s: str) -> str:
    """
    桶排序
    """
    counter = dict()
    max_freq = 0
    for i in s:
        counter[i] = counter.get(i, 0) + 1
        max_freq = max(max_freq, counter[i])

    # bucket sort
    buckets = [[] for _ in range(max_freq + 1)]
    for k, v in counter.items():
        buckets[v].append(k)

    ans = ''
    for i in range(max_freq, -1, -1):
        for ch in buckets[i]:
            ans += ch * i
    return ans


if __name__ == '__main__':
    r = frequency_sort_v2('tree')
    print(r)

File: algorithms/451-sort-characters-by-frequency/sort-characters-by-frequency.go
package problem451

import (
	"bytes"
	"sort"
)

// frequencySortV1 按照出现的频率排序
func frequencySortV1(s string) string {
	cnt := map[byte]int{}
	for i := range s {
		cnt[s[i]]++
	}

	type pair struct {
		ch  byte
		cnt int
	}
	pairs := make([]pair, 0, len(cnt))
	for k, v := range cnt {
		pairs = append(pairs, pair{k, v})
	}

	// sort by cnt
	sort.Slice(pairs, func(i, j int) bool {
		return pairs[j].cnt < pairs[i].cnt
	})

	ans := make([]byte, 0, len(s))
	for _, p := range pairs {
		ans = append(ans, bytes.Repeat([]byte{p.ch}, p.cnt)...)
	}
	return string(ans)
}

// frequencySortV2 桶排序
func frequencySortV2(s string) string {
	cnt := make(map[byte]int)
	maxFreq := 0
	for i := range s {
		cnt[s[i]]++
		maxFreq = max(maxFreq, cnt[s[i]])
	}

	buckets := make([][]byte, maxFreq+1)
	for ch, c := range cnt {
		buckets[c] = append(buckets[c], ch)
	}

	ans := make([]byte, 0, len(s))
	for i := maxFreq; i > 0; i-- {
		for _, ch := range buckets[i] {
			ans = append(ans, bytes.Repeat([]byte{ch}, i)...)
		}
	}

	return string(ans)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
